modernization of the pesuacademy library

8figalltimepro opened the pull request:
This PR begins the planned modernization of the `pesuacademy` library.

### Progress

- [x] **Phase 1:** Modernize data models with Pydantic
- [x] **Phase 2:** Add asynchronous page handlers for scraping
- [x] **Phase 3:** Replace sync client with modern async client

Commit:
feat: Replace old models with modern Pydantic models

Commit:
feat(pages): Add asynchronous page handlers for scraping

8figalltimepro commented:
**Changes in this Push (Phase 2):**

* Added the complete set of asynchronous page handlers to the `pesuacademy/pages/` directory.
* These handlers use `httpx` and `BeautifulSoup` to perform the scraping for each specific page (Courses, Units, Results, etc.).
* Added Google-style docstrings to the new handler methods.

Commit:
refactor(core): Replace main client with async facade

Commit:
refactor(core): Replace main client with async facade

8figalltimepro commented:
I've just pushed the final changes for the core library modernization. This PR now contains the complete refactor to an asynchronous, Pydantic-based library. I've marked it as 'Ready for review'. Thanks

Commit:
fix: Relative Imports

8figalltimepro commented:
Relative Import Fixes

Commit:
feat: fetch accurate profile information including base64 profile image

achyu-dev commented on pesuacademy/models/announcement.py:
So, all these fields you are fetching from an announcement page in PESU Academy?

8figalltimepro replied:
yes, there is a sep announcement page

achyu-dev commented on pesuacademy/pages/announcements.py:
you can below line to class-level constants - 

```python
base_url = "https://www.pesuacademy.com"
```

achyu-dev commented on pesuacademy/pages/attendance.py:
The request parameters can be made to a constants value, if its not changing at all. Instead of entering like this, looks at bit weird. 

Like this - 

```python
ATTENDANCE_MENU_ID = "660"
ATTENDANCE_CONTROLLER_MODE = "6407"
ATTENDANCE_ACTION_TYPE = "8"
```

achyu-dev commented on pesuacademy/pages/course_detail.py:
same here, magic numbers can be moved as constants

achyu-dev reviewed (changes requested):
Plz do below changes

Commit:
refactor(core): Centralize URL parameters into constants.py

aditeyabaral commented on pesuacademy/models/announcement.py:
```diff
@@ -1,27 +1,18 @@
+        content (str): The content of the announcement.
+        links (Optional[List[str]]): Optional list of attachment links related to the announcement.
+    """
+    title: str
+    date: datetime.date
+    content: str
+    links: Optional[List[str]] = None
```
Why did you remove image and files? I believe an announcement had an image and also additional links.

8figalltimepro replied:
PES is not using an image now
files is now handled by links as stated in the docs they are "download links to the attachment files"

aditeyabaral replied:
Do they not have images anymore as part of announcements? I can see images in the announcements when I log in

8figalltimepro replied:
no 😅

aditeyabaral replied:
<img width="1570" height="642" alt="image" src="https://github.com/user-attachments/assets/6b519eba-4bed-4695-8530-28649855762b" />
What about this

8figalltimepro replied:
🤨 these are part of some different set of announcement which only shows up at the dashboard, the don't show up in the announcements page.
same thing happened with the AT2025 announcement, it showed on the dashboard but not in the announcements page.
idk what PES thinks 😡

aditeyabaral replied:
I'd suggest adding support for image as well. I believe these will eventually end up in the announcement page (this is what used to happen a few years ago).

8figalltimepro replied:
https://github.com/pesu-dev/pesuacademy/issues/16

aditeyabaral commented on pesuacademy/models/announcement.py:
```diff
@@ -1,27 +1,18 @@
-        self.img = img
-        self.files = files
-
-    def __str__(self):
-        return f"{self.__dict__}"
```
I think you can replace all these `__str__` with the `rich` printing functions you had in your cli.

8figalltimepro replied:
https://github.com/pesu-dev/pesuacademy/issues/21

aditeyabaral commented on pesuacademy/models/course.py:
```diff
@@ -1,35 +1,33 @@
+        attendance (Optional[Attendance]): Attendance information for the course.
+        course_id (Optional[str]): Unique identifier for the course instance.
+    """
+    code: str
+    title: str
+    type: Optional[str] = None
+    status: Optional[str] = None
+    attendance: Optional[Attendance] = None
+    course_id: Optional[str] = None
```
Is `code` different from `ID`?

8figalltimepro replied:
yes code is UE24CS151B
ID would be the course id on the server side

I will improve the docs for it to better diff it by giving proper examples 👍

aditeyabaral replied:
Do we need both?

8figalltimepro replied:
yes, cause id will be used to by the course_detail.py to for a param field (e.g. , 19205) serves
code on the other hand will be used by the front-end effectively

aditeyabaral replied:
Hmm ok. Although not sure if it should be part of the model

aditeyabaral replied:
Keep both, rename `course_id` to `id`.

aditeyabaral commented on pesuacademy/models/profile.py:
```diff
@@ -1,144 +1,112 @@
+    permanent: str
+
+
+# Need to add Other details like profile picture, date of birth, Blood Group, etc. in the future.
```
Add "TODO:" to the comment, and raise an issue. Do not track TODO items via comments - always formalize them.

8figalltimepro replied:
👍 had to remove this comment as i added those but forgot to 😭

aditeyabaral commented on pesuacademy/models/course.py:
```diff
@@ -1,35 +1,33 @@
+        percentage (Optional[float]): Attendance percentage calculated from attended and total classes.
+    """
+    attended_classes: Optional[int] = None
+    total_classes: Optional[int] = None
```
Use `attended` and `total` as the attribute names

8figalltimepro replied:
@aditeyabaral 
please note that in the attendance.py, just using variables like total might not be the best way.
So i will keep them as attended_classes & total_classes but change it to attended & total in the model

aditeyabaral replied:
They will be part of the model schema, so someone will eventially use it as `course.attendance.total` and `course.attendance.attended`. Schema attribute names should always be relative to the parent. Please rename.

8figalltimepro replied:
done

aditeyabaral commented on pesuacademy/models/materials.py:
```diff
@@ -0,0 +1,36 @@
+        unit_id (str): Unique identifier for the unit.
+    """
+    title: str
+    unit_id: str
```
Rename to `id`

8figalltimepro replied:
seems a bit ambiguous as in the Topic class there are 2 diff ids so better to keep it separate at all places, unit_id is better kept same in both places

aditeyabaral replied:
Same as earlier: schema attributes are named relative to the parent model/class. Which means `id` here would refer to the unit. This allows doing something like `unit.id == topic.unit_id`

8figalltimepro replied:
Done

8figalltimepro replied:
https://github.com/pesu-dev/pesuacademy/issues/22

aditeyabaral commented on pesuacademy/models/materials.py:
```diff
@@ -0,0 +1,36 @@
+    unit_id: str
+
+
+class Topic(BaseModel):
+    """ Represents a topic within a unit in the PESU Academy system.
+    Attributes:
+        title (str): The title of the topic.
+        topic_id (str): Unique identifier for the topic.
+        course_id (str): Identifier for the course this topic belongs to.
+        unit_id (str): Identifier for the unit this topic belongs to.
+    """
+    title: str
+    topic_id: str
+    course_id: str     
+    unit_id: str    
```
If a unit can have multiple topics, why not add an attribute called `topics` in the Unit model, and have that store a list of topics?

8figalltimepro replied:
The current design intentionally separates fetching units from fetching topics to make the library more efficient and flexible. A user might only want the list of unit titles, and forcing an extra network call for every topic in that case would be inefficient. This approach gives the developer using the library the control to fetch deeper information only when they need it.

aditeyabaral replied:
Are they not part of the same page? If they need an extra network call, then there is no need. If they are part of the same network call (basically rendered on the same page), then add them. Otherwise, good design principle.

8figalltimepro replied:
sadly, they are not part of the same page and requires separate  network call

aditeyabaral replied:
Then no worries. Keep it as it is

aditeyabaral commented on pesuacademy/models/profile.py:
```diff
@@ -1,144 +1,112 @@
+        email_id (str): Email address of the user.
+        contact_no (str): Contact number of the user.
+        name_as_in_aadhar (str): Name as per Aadhar card.
+        profile_image_base64 (Optional[str]): Base64 encoded profile image, if available.
```
Rename to `image`

aditeyabaral commented on pesuacademy/models/profile.py:
```diff
@@ -1,144 +1,112 @@
+    sslc_marks: str
+    puc_marks: str
+    date_of_birth: str
+    blood_group: str
```
Nitpick, but some of these are known constants. For attributes that will always be one among a set of known deterministic values, we should use a `Literal` type. This applies to all such attributes across other models too.

aditeyabaral replied:
Resolve this for now since it isn't an important field, but raise an issue to document future work.

8figalltimepro replied:
https://github.com/pesu-dev/pesuacademy/issues/23

aditeyabaral commented on pesuacademy/models/results.py:
```diff
@@ -0,0 +1,46 @@
+    max_marks: Optional[str] = None
+
+
+class SubjectResult(BaseModel):
```
Rename to `CourseResult` for clarity

aditeyabaral commented on pesuacademy/models/results.py:
```diff
@@ -0,0 +1,46 @@
+    """
+    name: str
+    marks: Optional[str] = None
+    max_marks: Optional[str] = None
```
`max_marks` can be ambiguous: is this the total marks (100), or the max someone obtained (95)?

8figalltimepro replied:
max_marks is the maximum marks one can score in an assessment.
max_marks ~ total_marks
i can change it

8figalltimepro replied:
changed it to total_marks everywhere

aditeyabaral replied:
What about renaming this to `total`?

8figalltimepro replied:
will do it 👍

8figalltimepro replied:
done

aditeyabaral commented on pesuacademy/models/seating_information.py:
```diff
@@ -1,19 +1,19 @@
+        block (str): Block or section of the seating arrangement.
+    """
+    name: str
+    course_code: str
```
Open discussion; does it make sense to have a course field that stores the Course model from course.py? Otherwise, to fetch details of a course from the seating information, someone will have to fetch the IDs, and then from those IDs list *all* courses, and then filter using that. However doing so will also increase network calls.

8figalltimepro replied:
The current design prioritizes a fast initial response by not making extra network calls. It provides the course_code so a developer can choose to fetch more details if they need them. Eagerly fetching the full Course object for every subject in the seating plan would significantly slow down the get_seating_info() call.

aditeyabaral replied:
Makes sense. Good point.

aditeyabaral commented on pesuacademy/pages/announcements.py:
```diff
@@ -1,124 +1,88 @@
-from pesuacademy.models.announcement import Announcement, AnnouncementFile
+from typing import List
+from ..models.announcement import Announcement
+from .. import constants
```
Follow good important standards

aditeyabaral commented on pesuacademy/pages/announcements.py:
```diff
@@ -1,124 +1,88 @@
         soup = BeautifulSoup(response.text, "lxml")
+        
+        announcements = []
+        base_url = "https://www.pesuacademy.com"
```
I think we can move this to constants

8figalltimepro replied:
this is not a constant per se, just used to form the final download link.

8figalltimepro replied:
but i will still put it in the constants

aditeyabaral replied:
Move all hardcoded strings/numbers/values to a constants file. These are known as "magic values" in code because there is no way this was derived apart from hard-coding them. These values should *all* be reserved in a separate file(s)

8figalltimepro replied:
done

aditeyabaral commented on pesuacademy/pages/announcements.py:
```diff
@@ -1,124 +1,88 @@
+                announcements.append(
+                    Announcement(title=title, date=date, content=content, links=links or None)
+                )
 
-    def get_page(
-        self,
-        session: requests_html.HTMLSession,
-        csrf_token: str,
-        start_date: Optional[datetime.date] = None,
-        end_date: Optional[datetime.date] = None,
-    ) -> list[Announcement]:
-        url = "https://www.pesuacademy.com/Academy/s/studentProfilePESUAdmin"
-        query = {
-            "menuId": "667",
-            "controllerMode": "6411",
-            "actionType": "5",
-            "_": str(int(datetime.datetime.now().timestamp() * 1000)),
-        }
-        response = session.get(url, allow_redirects=False, params=query)
-        if response.status_code != 200:
-            raise ConnectionError("Unable to fetch announcement data.")
-        soup = BeautifulSoup(response.text, "lxml")
-
-        announcement_ids = soup.find_all("a", class_="pull-right readmorelink")
-        pattern = re.compile(r"handleShowMoreAnnouncement\(\d+, \d+,(\d+)\)")
-        announcement_ids = [
-            pattern.match(ann.attrs.get("onclick")).group(1) for ann in announcement_ids
-        ]
-
-        announcements = list()
-        for announcement_id in announcement_ids:
-            announcement = self.get_announcement_by_id(
-                session, csrf_token, announcement_id
-            )
-            if start_date and announcement.date < start_date:
-                continue
-            if end_date and announcement.date > end_date:
+            except (AttributeError, ValueError) as e:
+                # Skip any panels that have parsing errors
+                print(f"Skipping a panel due to parsing error: {e}")
```
Parallelise this. Move this to a function and spawn a background thread to fetch and parse details. Check the auth repo for a reference.

8figalltimepro replied:
In our scraper, the loop that parses the announcement HTML is a CPU-bound task. I've found that the entire parsing process is extremely fast, it's measured in milliseconds. The real bottleneck by far is the initial network call to download the page itself.

Adding multiprocessing for that specific loop would offer a negligible real case speed improvement. We've focused our parallelization efforts on the slow network calls using asyncio.gather (like when fetching all material links or all courses), as that's where the most significant performance gains are

In the benchmark.py scrip you are parallelizing an I/O-bound task. It uses threads to make many slow network requests concurrently, which provides a huge performance gain but not so in our case

aditeyabaral commented on pesuacademy/pages/attendance.py:
```diff
@@ -1,71 +1,66 @@
+                    code=cols[0], title=cols[1], attendance=course_attendance
+                )
+                attendance_data.append(course)
+        return attendance_data
```
Same here. Parallelize wherever possible to speed up processing.

aditeyabaral commented on pesuacademy/pages/course_detail.py:
```diff
@@ -0,0 +1,58 @@
+                unit_id = match.group(1)
+                units.append(Unit(title=title, unit_id=unit_id))
+        
+        return units
```
Evaluate whether this needs parallelization.

aditeyabaral commented on pesuacademy/pages/courses.py:
```diff
@@ -1,60 +1,62 @@
+                        course_id=course_id
+                    )
+                )
+        return courses
```
Same here.

aditeyabaral commented on pesuacademy/pages/courses.py:
```diff
@@ -1,60 +1,62 @@
-        return courses
+            cols = [c.text.strip() for c in row.find_all("td")]
+
+            if len(cols) >= 4: # Temp fix to ensure we have enough columns
```
What does this mean?

8figalltimepro replied:
I am just making sure that the format of the page is what I expect it to be

aditeyabaral replied:
Document this in the code so it is clearer. Use `assert` to gracefully catch these instead of `if`s

8figalltimepro replied:
https://github.com/pesu-dev/pesuacademy/issues/25

aditeyabaral commented on pesuacademy/pages/esaresult.py:
```diff
@@ -0,0 +1,113 @@
+        # Fetch the SGPA
+        sgpa_raw = summary_divs[1].contents[-1].strip()
+        # Not fetching CGPA as it is not a part of ESA results for a particular semester
+        # Will Fetch it separately if needed later
```
Could you clarify this? Why is the CGPA on the SGPA page?

8figalltimepro replied:
The esa result page has a CGPA field which is the cgpa of the student when he/she was in that semester
A student studying in Sem-6, if he/she checks their esa_result of sem-2 their cgpa will be shown as the cgpa he/she had in sem-2, not the current one. I felt this was useless info. 
There is a better way to get the current CGPA, which I will be adding in later features

aditeyabaral replied:
I think it would be good to fetch it since it is on the same page. Sure, its not relevant to the page, but the goal is to provide read-only access to *everything* on the page. We should not filter/morph any details.

8figalltimepro replied:
Let's discuss this in an alternate issue because it will be unnecessarily confusing if it's kept as CGPA or SEM_CGPA. Best way would be to make it private?; There is a lot to discuss here so best to keep it in  a separate issue
https://github.com/pesu-dev/pesuacademy/issues/27

aditeyabaral commented on pesuacademy/pages/esaresult.py:
```diff
@@ -0,0 +1,113 @@
+                code=code, title=title, 
+                credits_earned=s_credits_earned, credits_total=s_credits_total,
+                assessments=assessments
+            ))
```
Same comment - parallelize

aditeyabaral commented on pesuacademy/pages/esaresult.py:
Rename the file. You should never combine words without an `_` for filenames/variables

aditeyabaral commented on pesuacademy/pages/material_links.py:
```diff
@@ -0,0 +1,69 @@
+            "unitid": topic.topic_id,
+            "menuId": constants.PageURLParams.MaterialLinks.MENU_ID,
+            "_": str(int(datetime.datetime.now().timestamp() * 1000)),
+        }
```
Instead of declaring this dictionary multiple times, can we create a helper function that takes in the keys as args and returns this dictionary? It simplifies the code in almost all files

aditeyabaral replied:
The same helper function can be called by any page

8figalltimepro replied:
Good point. But since the parameters for each request are slightly different, a single helper function could get complicated. A better approach might be to define a DEFAULT_PARAMS dictionary in constants.py and have each handler function merge it with its own specific parameters. I'll look into implementing this.

8figalltimepro replied:
Ok, That didnt work as I expected 😅
Will revert back to your solution

8figalltimepro replied:
Solved

aditeyabaral commented on pesuacademy/pages/material_links.py:
```diff
@@ -0,0 +1,69 @@
+                        # Construct the full download URL
+                        full_url = f"{base_url}{partial_url}"
+                        links.append(MaterialLink(title=title, url=full_url, is_pdf=True))
+
```
Same here

aditeyabaral commented on pesuacademy/pages/esaresult.py:
```diff
@@ -0,0 +1,113 @@
+
+class ResultsPageHandler:
+    @staticmethod
+    async def get_page(session: httpx.AsyncClient, semester_id: str) -> SemesterResult:
```
If these functions are not meant to be called by the user, make them private. Comment applies to all such functions

8figalltimepro replied:
yes, privated everything necessary 👍

aditeyabaral commented on pesuacademy/pages/profile.py:
```diff
@@ -1,32 +1,146 @@
+        )
+
+    @staticmethod
+    async def get_page(session: httpx.AsyncClient) -> Profile:
```
Maybe we can rename this interface to `_get` for all pages

8figalltimepro replied:
Have kept it _get_page for now, let's revisit this another time, as it's just called internally, isn't of high priority

aditeyabaral replied:
Document this in an issue.

aditeyabaral replied:
You can resolve this after raising the issue.

8figalltimepro replied:
https://github.com/pesu-dev/pesuacademy/issues/28

aditeyabaral commented on pesuacademy/pages/seating_information.py:
```diff
@@ -1,58 +1,49 @@
-                return SeatingInformationHandler.get_seating_information_from_page(soup)
-        except Exception:
-            raise ConnectionError("Unable to fetch seating info.")
+        soup = BeautifulSoup(response.text, "lxml")
```
Rewrite all bs4 syntax with selectolax

aditeyabaral commented on pesuacademy/pages/seating_information.py:
```diff
@@ -1,58 +1,49 @@
+        seating_info = []
+        for row in info_table.find("tbody").find_all("tr"):
+            cols = [c.text.strip() for c in row.find_all("td")]
+            if len(cols) >= 6:
```
Where are these magic numbers coming from? You should define them in constants

8figalltimepro replied:
https://github.com/pesu-dev/pesuacademy/issues/25

aditeyabaral commented on pesuacademy/pages/unit.py:
```diff
@@ -0,0 +1,69 @@
+                    course_id=course_id,
+                    unit_id=scraped_unit_id
+                )
+            )
```
Same here

aditeyabaral commented on pesuacademy/client.py:
```diff
@@ -0,0 +1,116 @@
+        self._session = httpx.AsyncClient(base_url=self._base_url, follow_redirects=True, timeout=30.0)
+        self._csrf_token: Optional[str] = None
+        self._semester_ids: Dict[int, str] = {}
+        self.is_authenticated = False
```
This is redundant. This library requires you to log in to use it.

8figalltimepro replied:
Will be removed 👍

aditeyabaral commented on pesuacademy/constants.py:
```diff
@@ -0,0 +1,50 @@
+class PageURLParams:
```
Define this as a [dataclass](https://www.dataquest.io/blog/how-to-use-python-data-classes/).

8figalltimepro replied:
done

aditeyabaral commented on pesuacademy/pesuacademy.py:
```diff
@@ -1,186 +1,176 @@
+        Args:
+            client (PesuAcademyClient): An authenticated instance of the core client.
+        """
+        self._client = client
```
The `PESUAcademy` class _is_ the client, which serves as the interface between the user and the PESU server. Its also ambiguous to have 2 very similar class names: `PESUAcademy` and `PESUAcademyClient`. 

Rename this to `handler` or something else (discuss here). Or, if you want to retain the `_client` attribute, rename the class to something else.

aditeyabaral replied:
I think keeping it as `client` is fine, but you should rename the class.

8figalltimepro replied:
done, renamed

aditeyabaral commented on pesuacademy/pesuacademy.py:
```diff
@@ -1,186 +1,176 @@
+        self._client = client
+
+    @classmethod
+    async def login(cls, username: str, password: str) -> "PESUAcademy":
+        """
+        Creates, authenticates, and returns an active PESUAcademy session.
+        This is the primary way to start using the library.
+
+        Args:
+            username (str): The user's SRN, PRN, or other login identifier.
+            password (str): The user's password.
+
+        Returns:
+            PESUAcademy: An authenticated instance ready to make requests.
+        """
+        client = PesuAcademyClient()
+        await client.login(username, password)
+        return cls(client)
```
This function should allow logging in without passing arguments by reading from the environments. Credentials are usually read from the env in production systems, not passed as arguments.

aditeyabaral commented on pesuacademy/client.py:
```diff
@@ -0,0 +1,116 @@
+        self._semester_ids = await SemesterHandler.get_semester_ids(self._session)
+
+    async def get_seating_info(self) -> List[SeatingInformation]:
+        if not self.is_authenticated: raise Exception("Not authenticated.")
```
This `if` is dead code. They cannot create the parent object without logging in

aditeyabaral commented on pesuacademy/client.py:
This file seems redundant. Why not move everything to PESUAcademy? Most of the `get_*` functions just call another function.

8figalltimepro replied:
already discussed

aditeyabaral reviewed (changes requested):
General comments: 

1. Use ruff for formatting
2. Use python 3.11+ syntax style (Replace `Optional`, `List` etc with modern equivalents)

Commit:
refactor: Implement major feedback from PR review

8figalltimepro commented:
Hi @aditeyabaral , thank you for the detailed feedback.  
I have just pushed a new commit that addresses most of the points and refactors the library to be cleaner and more robust.

---

## Implemented Changes

<details>
<summary>Click to see all implemented changes</summary>

### Model Naming & Structure
- Renamed `profile_image_base64` → `image`.
- Renamed `max_marks` → `total_marks` for better clarity.
- Renamed `SubjectResult` → `CourseResult` for consistency.
- Renamed the core `PesuAcademyClient` → `_PesuScraper` to mark it as internal and avoid confusion with the public `PESUAcademy` facade.

### Code Structure & Readability
- Moved `base_url` to a central `constants.py` file.
- Created a `util/utils.py` file with a `build_params` helper to eliminate repetitive parameter dictionary creation in the page handlers.
- Privatized all page handler classes and methods (e.g., `_CoursesPageHandler._get_page`) to enforce the public facade pattern and hide internal implementation.

### Credentials Management
- `PESUAcademy.login` now supports reading credentials from `.env` files / environment variables if not provided as arguments.

</details>

---

## New Feature: I/O Parallelization
- Methods for fetching **Courses** and **Attendance** for all semesters now run their network calls in parallel using `asyncio.gather()`.  
- This significantly speeds up the process when a specific semester is not provided.

---

## Discussion on Other Points

**On Eagerly Loading Data (e.g., adding topics to the Unit model):**  
The design keeps fetching units and topics separate for efficiency. Many users may only want unit titles, and forcing extra network calls would be wasteful. This gives developers flexibility to fetch deeper information only when needed.

**On Parallelizing Parsing Loops:**  
Benchmarks show that parsing is extremely fast (milliseconds). Multiprocessing overhead made it slower, so the real performance gains come from the newly added I/O parallelization.

**On Deferring Announcement Image Scraping:**  
Image scraping for announcements is deferred to a future release. Current test data lacks examples, so reliable parsing logic cannot yet be implemented.

**On `is_authenticated` Checks:**  
Although the login flow already prevents unauthenticated access, keeping the check adds a defensive layer. It ensures future code changes will not allow unauthenticated calls, improving robustness.

---

achyu-dev committed:
added precommit hooks

achyu-dev committed:
removed black

achyu-dev committed:
added system dependencies to lint

achyu-dev committed:
added system dependencies to build docs

achyu-dev committed:
added system dependencies to pre-commit

aditeyabaral committed:
Add dev stuff

aditeyabaral commented on setup.py:
```diff
@@ -1,40 +0,0 @@
-    long_description = DESCRIPTION
-
-setuptools.setup(
-    name="pesuacademy",
-    version=VERSION,
-    author="Aditeya Baral",
-    author_email="aditeya.baral@gmail.com",
-    description=DESCRIPTION,
-    long_description=long_description,
-    long_description_content_type="text/markdown",
-    url="https://github.com/HackerSpace-PESU/pesuacademy-py",
-    packages=setuptools.find_packages(),
-    install_requires=[
-        "requests",
-        "requests-html",
-        "beautifulsoup4",
-        "lxml_html_clean"
-    ],
-    classifiers=[
-        "Development Status :: 5 - Production/Stable",
-        "Intended Audience :: Developers",
-        "Programming Language :: Python :: 3.9",
-        "Programming Language :: Python :: 3.10",
-        "Programming Language :: Python :: 3.11",
-        "License :: OSI Approved :: MIT License",
-        "Operating System :: OS Independent",
-    ],
-    python_requires='>=3.9',
-    keywords=["pesu", "pesu academy", "api", "wrapper", "python"]
-)
```
We will need to eventually migrate some of these to `pyproject.toml`

aditeyabaral commented on python:
What is this file? Delete.

aditeyabaral commented on pyproject.toml:
```diff
@@ -0,0 +1,71 @@
+requires-python = ">=3.12"
+license = { text = "MIT" }
+authors = [
+    { name = "Aditeya Baral", email = "aditeya.baral@gmail.com" }
```
Add @8figalltimepro, @abaksy's names and emails here.

8figalltimepro replied:
Done for mine

aditeyabaral commented on materials.csv:
What is this file? Delete

aditeyabaral commented on LICENSE:
```diff
@@ -1,6 +1,6 @@
 MIT License
 
-Copyright (c) 2024 HackerSpace @PESU
+Copyright (c) 2025 Aditeya Baral
```
I think we should move the copyrights to PESU Dev. @sach-12?

sach-12 replied:
No, leave it to either you or me. PESU Dev is not a legally recognized entity to use for licensing

aditeyabaral commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,30 @@
+      - id: name-tests-test
+        args: [ '--pytest-test-first' ]
+
+  # - repo: local
+  #   hooks:
+  #     - id: pytest
+  #       name: Run Pytest
+  #       entry: python scripts/run_tests.py
+  #       language: system
+  #       types: [ python ]
+  #       pass_filenames: false
```
We will revisit this later when we add tests for our package. Can someone delete this?

aditeyabaral commented on .env.example:
We will need these for tests, so keeping them here.

aditeyabaral commented on .github/ISSUE_TEMPLATE/bug_report.yaml:
@achyu-dev @ndigvijay please update this template accordingly

aditeyabaral commented on .github/ISSUE_TEMPLATE/config.yml:
```diff
@@ -0,0 +1,13 @@
+
+  - name: 💬 Join the PESU Developer Group Discord Server
+    url: https://discord.gg/eZ3uFs2
+    about: Join our Discord for help with PESUAuth integration, support, or bot development.
```
Update the sentence

aditeyabaral commented on .github/ISSUE_TEMPLATE/feature_request.yaml:
@achyu-dev @ndigvijay please update this template accordingly

aditeyabaral commented on .github/workflows/build-docs.yml:
```diff
@@ -11,6 +11,10 @@ jobs:
     steps:
       - uses: actions/checkout@v3
       - uses: actions/setup-python@v3
+      - name: Install system dependencies for lxml
+        run: |
+          sudo apt-get update
+          sudo apt-get install -y libxml2-dev libxslt1-dev zlib1g-dev
```
Do we need this?

achyu-dev replied:
yes, because as told previously, need this for `setuptools` or some other library.

aditeyabaral commented on .github/workflows/deploy.yaml:
This file can be deleted, we do not have a docker image here.

achyu-dev replied:
Eventually we need to add anyway, for now sure we will delete it

aditeyabaral commented on .github/CODEOWNERS:
```diff
@@ -0,0 +1,23 @@
+# Maintainers for all files in the repository
+* @pesu-dev/auth-maintainers
```
Update this to the right team

aditeyabaral commented on .github/CODEOWNERS:
```diff
@@ -0,0 +1,23 @@
+# Repository owner review for critical files
+
+## GitHub Configuration and Templates
+.github/ @aditeyabaral
+
+## Project Configuration
+pyproject.toml @aditeyabaral
+requirements.txt @aditeyabaral
+.pre-commit-config.yaml @aditeyabaral
+.env.example @aditeyabaral
+
+## Documentation and Legal
+README.md @aditeyabaral
+LICENSE @aditeyabaral
+
+## Infrastructure and Deployment
+Dockerfile @aditeyabaral
+
+## Testing
+tests/ @aditeyabaral
```
Update this accordinly. Add me and @8figalltimepro as owner reviewers for critical files (you will need to update which files are "critical"). Also add @abaksy for all packaging related files.

aditeyabaral commented on .github/CONTRIBUTING.md:
@achyu-dev @ndigvijay please update this template accordingly

aditeyabaral commented on .github/PULL_REQUEST_TEMPLATE.md:
@achyu-dev @ndigvijay please update this template accordingly

aditeyabaral commented on .github/SECURITY.md:
@achyu-dev @ndigvijay please update this template accordingly

aditeyabaral commented on docs/conf.py:
@abaksy can you review if these files are still needed with `uv`?

aditeyabaral commented on docs/index.rst:
@abaksy can you review if these files are still needed with `uv`? If yes, update accordingly

aditeyabaral commented on pesuacademy.egg-info/dependency_links.txt:
Delete this file, folder and add to gitignore

aditeyabaral commented on pesuacademy.egg-info/PKG-INFO:
Delete this file, folder and add to gitignore

aditeyabaral commented on pesuacademy.egg-info/requires.txt:
Delete this file, folder and add to gitignore

aditeyabaral commented on pesuacademy.egg-info/SOURCES.txt:
Delete this file, folder and add to gitignore

aditeyabaral commented on pesuacademy.egg-info/top_level.txt:
Delete this file, folder and add to gitignore

aditeyabaral commented on pesuacademy/models/announcement.py:
```diff
@@ -1,27 +1,19 @@
+        title (str): The title of the announcement.
+        date (datetime.date): The date of the announcement.
+        content (str): The content of the announcement.
+        links (Optional[List[str]]): Optional list of attachment links related to the announcement.
```
It would make more sense to call the field `attachments`

aditeyabaral commented on pesuacademy/models/course.py:
```diff
@@ -1,35 +1,35 @@
+    type: str | None = None
+    status: str | None = None
+    attendance: Attendance | None = None
+    course_id: str | None = None
```
Ideally, this is something the library would use more than the user, right? Its not something the user would concern with since this is used in the network calls I believe. In that case, please rename this to `_id`. We want it to be private.

8figalltimepro replied:
But in this use case, the user actually needs it (because get_units_for_course(course.id) requires course.id).

If you rename it to _id, you’re signaling “don’t touch this,” but at the same time, the user will have no obvious way to pass the right argument.

This goes same for other places where the id is being refactored - The topics having id can be used by the user to store them and as a unique key.

aditeyabaral commented on pesuacademy/models/materials.py:
```diff
@@ -0,0 +1,43 @@
+    """
+
+    title: str
+    topic_id: str
```
Should be `id`

8figalltimepro replied:
https://github.com/pesu-dev/pesuacademy/issues/22

aditeyabaral commented on pesuacademy/models/results.py:
```diff
@@ -0,0 +1,50 @@
+    code: str
+    title: str
+    credits_earned: str
+    credits_total: str
```
I thnk it would make more sense to have a model for this, because we created a similar model for Parent details. Something like `Credits` and it contains two attributes: `earned` and `total`.  Also cause these 2 fields are reused below in Sem Result also.

8figalltimepro replied:
Done

aditeyabaral commented on pesuacademy/pages/announcements.py:
```diff
@@ -1,124 +1,89 @@
 
-from pesuacademy.models.announcement import Announcement, AnnouncementFile
+from .. import constants
+from ..models.announcement import Announcement
```
Please use good import etiquette. Import from the root level always.

8figalltimepro replied:
done

aditeyabaral commented on pesuacademy/pages/announcements.py:
```diff
@@ -1,124 +1,89 @@
+            httpx.HTTPStatusError: If the request to the announcements page fails.
+        """
+        url = "/s/studentProfilePESUAdmin"
+        params = build_params(constants.PageURLParams.Announcements, url="studentProfilePESUAdmin")
```
Make `build_params` private.

8figalltimepro replied:
done

aditeyabaral commented on pesuacademy/pages/material_links.py:
```diff
@@ -0,0 +1,72 @@
+
+from .. import constants
+from ..models.materials import MaterialLink, Topic
+from ..util import build_params
```
Fix these in all corresponding files

8figalltimepro replied:
done

aditeyabaral commented on pesuacademy/pages/semester.py:
```diff
@@ -0,0 +1,57 @@
+
+        semester_ids = {}
+
+        # Too many fixes done, need to refactor this later
```
Add some more clarity here

8figalltimepro replied:
removed that comment
There was multiple fixes I had to do where in the semid extracted  from the page had weird formatting text with it, so had to clean it up (written that then)

aditeyabaral commented on pesuacademy/util/utils.py:
```diff
@@ -0,0 +1,25 @@
+from typing import Any
+
+
+def build_params(params_obj: Any, **kwargs: Any) -> dict[str, str]:
```
Not user facing, rename to `_build_params`

8figalltimepro replied:
done

aditeyabaral commented on pesuacademy/client.py:
```diff
@@ -0,0 +1,150 @@
+
+    async def get_material_links(self, topic: Topic, material_type_id: str) -> list[MaterialLink]:
+        if not self.is_authenticated:
+            raise Exception("Not authenticated.")
```
Add an issue for this later: we should use custom exceptions.

aditeyabaral commented on pesuacademy/constants.py:
Should be a private module

aditeyabaral reviewed (commented):
> On is_authenticated Checks:
Although the login flow already prevents unauthenticated access, keeping the check adds a defensive layer. It ensures future code changes will not allow unauthenticated calls, improving robustness.

While this sounds good, it is a little flawed given the context of our implementation. The variable is explicitly set by us, not by the user. Additionally, the user cannot make any function calls to the functions that require authentication, since the parent class itself will not initialise without authentication. This `if` condition while being good, does practically nothing because the client, if initialized, will _always_ be authenticated. If it is not authenticated, it means the client does not exist at all -> functions cannot be called.

8figalltimepro committed:
Delete python

Don't know what file this is but deleting it

achyu-dev committed:
fixed formatting errors

Commit:
refactor: Minor changes

8figalltimepro commented:
I have pushed a commit based on these changes
refactor: model and client updates

- Announcement model: renamed `links` → `attachments`
- Course model: renamed `course_id` → `id`
- Materials model: renamed `unit_id` → `id`, `topic_id` → `id`
- Results: introduced new `Credits` model with `earned` and `total`
- Updated relative imports → absolute imports for safer, clearer usage
- Privatized `_build_params` for internal-only access
- Client: removed redundant `self.is_authenticated` checks
- Privatized `_PageURLParams` for internal handling

aditeyabaral commented on pesuacademy/client.py:
```diff
@@ -0,0 +1,124 @@
+        self._session = httpx.AsyncClient(base_url=self._base_url, follow_redirects=True, timeout=30.0)
+        self._csrf_token: str | None = None
+        self._semester_ids: dict[int, str] = {}
+        self.is_authenticated = False
```
Remove this if it isn't being used anymore

8figalltimepro replied:
Will be removed 👍

silicoflare reviewed (dismissed):
LGTM. Please finish merging this PR as soon as possible, because it is blocking other projects.

aditeyabaral commented on .github/CONTRIBUTING.md:
Hasn't been updated yet. Please update this.

aditeyabaral commented on .github/ISSUE_TEMPLATE/bug_report.yaml:
Same for this

aditeyabaral commented on .github/ISSUE_TEMPLATE/config.yml:
Same here.

aditeyabaral commented on .github/ISSUE_TEMPLATE/feature_request.yaml:
Same

aditeyabaral commented on .github/PULL_REQUEST_TEMPLATE.md:
Same

aditeyabaral commented on .github/SECURITY.md:
Same

aditeyabaral commented on pesuacademy/pages/semester.py:
```diff
@@ -0,0 +1,57 @@
+            text_match = re.search(r"\d+", text)
+
+            if raw_value and text_match:
+                # Imp Fix: Clean the raw value to extract the numeric part
```
If this is a TODO, raise an issue. Do not track TODO items in comments

8figalltimepro replied:
Not a TODO
cleaned it up to keep it like a normal comment

aditeyabaral commented on pesuacademy/pages/announcements.py:
```diff
@@ -1,124 +1,94 @@
+        Raises:
+            httpx.HTTPStatusError: If the request to the announcements page fails.
+        """
+        url = "/s/studentProfilePESUAdmin"
```
Isn't this URL common across all modules? Should we move it to constants?

8figalltimepro replied:
yes except the semester-url
Will do this fix 👍 (had planned to do it but forgot about it)

aditeyabaral commented on pesuacademy/models/__init__.py:
```diff
@@ -1,13 +1,37 @@
+"""This module initializes the models for the PESU Academy application."""
```
> PESU Academy application

This is not an application, this is a package. Make this change across all such occurrences.

8figalltimepro replied:
Done 👍

Commit:
refactor: Minor changes

aditeyabaral commented on pesuacademy/pages/__init__.py:
```diff
@@ -1,5 +1,25 @@
-from .courses import CoursesPageHandler
-from .profile import ProfilePageHandler
-from .seating_information import SeatingInformationHandler
+"""This module initializes the page handlers for the PESU Academy application."""
```
Rename application to package.

8figalltimepro replied:
Please Recheck the latest commit

aditeyabaral replied:
My bad, resolving

achyu-dev committed:
removed pesuacademy-egginfo

achyu-dev reviewed (approved):
LGTM ! Great work @8figalltimepro 
Approved 👍
